diff --git a/Arduino/System/BoardToolchain.cmake b/Arduino/System/BoardToolchain.cmake
index a1cb266..541a119 100644
--- a/Arduino/System/BoardToolchain.cmake
+++ b/Arduino/System/BoardToolchain.cmake
@@ -170,6 +170,19 @@ function (SetupBoardToolchain boards_namespace board_id generate_dir)
 	properties_set_value("ard_global" "build.system.path"
 		"${_core_pl_path}/system")
 
+	# Set compiler.sdk.path and ARDUINO_BOARD_COMPILER_SDK_PATH
+	_board_get_property("build.mcu" _prop_value QUIET DEFAULT "")
+	if (_prop_value)
+		_board_get_ref_platform("${_prop_value}" _mcu_pkg_name _build_mcu)
+		set(ARDUINO_BOARD_COMPILER_SDK_PATH "${ARDUINO_BOARD_RUNTIME_PLATFORM_PATH}/tools/sdk/${_build_mcu}")
+	else()
+		set(ARDUINO_BOARD_COMPILER_SDK_PATH)
+	endif()
+
+	properties_set_value("ard_global" "compiler.sdk.path"
+		"${ARDUINO_BOARD_COMPILER_SDK_PATH}")
+	string_escape_quoting(ARDUINO_BOARD_COMPILER_SDK_PATH)
+
 	# Set runtime.os and ARDUINO_BOARD_HOST_NAME
 	if (${CMAKE_HOST_APPLE})
 		set(ARDUINO_BOARD_HOST_NAME "macosx")
diff --git a/Arduino/Templates/ArduinoSystem.cmake.in b/Arduino/Templates/ArduinoSystem.cmake.in
index 73d974c..07d1c79 100644
--- a/Arduino/Templates/ArduinoSystem.cmake.in
+++ b/Arduino/Templates/ArduinoSystem.cmake.in
@@ -20,6 +20,7 @@ set(ARDUINO_BOARD_RUNTIME_PLATFORM_PATH "@ARDUINO_BOARD_RUNTIME_PLATFORM_PATH@")
 set(ARDUINO_CORE_SPECIFIC_PLATFORM_PATH "@ARDUINO_CORE_SPECIFIC_PLATFORM_PATH@")
 set(ARDUINO_BOARD_BUILD_CORE_PATH "@ARDUINO_BOARD_BUILD_CORE_PATH@")
 set(ARDUINO_BOARD_BUILD_VARIANT_PATH "@ARDUINO_BOARD_BUILD_VARIANT_PATH@")
+set(ARDUINO_BOARD_COMPILER_SDK_PATH "@ARDUINO_BOARD_COMPILER_SDK_PATH@")
 set(ARDUINO_BOARD_HOST_NAME "@ARDUINO_BOARD_HOST_NAME@")
 
 @ARDUINO_SEL_MENU_SET_LIST@
